ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs

```diff
@@ -492,6 +494,14 @@ addBlockToAdd tracer (BlocksToAdd queue) punish blk = do
 getBlockToAdd :: IOLike m => BlocksToAdd m blk -> m (BlockToAdd m blk)
 getBlockToAdd (BlocksToAdd queue) = atomically $ readTBQueue queue
 
+-- | Flush the 'BlocksToAdd' queue and notify the waiting threads.
+--
+closeBlocksToAdd :: IOLike m => BlocksToAdd m blk -> STM m ()
```

Comment:
Bah; this doesn't seem enough. We have a race now, don't we?

- Some threads are adding tasks to the queue; these are BlockFetch clients and the `NodeKernel.hs` forge.
- One thread is popping one task from the queue at a time

If the popping thread is killed mid-pop, then it notifies the unlucky owner of the task that got interrupted. And it also flushes the queue, similarly notifying all other task owners. And then the popping thread will terminate, since `bracket*` re-raises the exception.

But there's no guarantee the popping thread will be the last to terminate. So even when it's gone, other threads may be adding to the queue.

The first option that comes to mind:

- ~Complicate the queue by adding state indicating whether it's "open" or "closed" and have the `addBlockRunner` close the queue when it flushes it.~ Hmm... the `ChainDB` already has an open/closed state; maybe the `addBlockRunner` dying is either reason enough to fully close the ChainDB or can only actually happen when the ChainDB is closed/closing or something like that?

- ~But now `addBlockAsync` will _also_ be partial, since you can't add to a closed queue.~ Now `addBlockAsync :: ... -> m (AddBlockPromise m blk)` would create a degenerate "promise" (immediately filled with `False` and `Nothing`) when asked to add a block to a closed queue.

Reply:
I didn't really grasp the problem here. Isn't the point of STM to be atomic, I don't think we'll get a Async exception mid-pop, but what can happen is the thread getting interrupted while blocked on waiting for popping. And if that's the case the cleanup handle won't even run.

Reply:
@nfrisby is right, we need to make sure no new block can be accepted by the db after `closeBlocksToAdd` is called by `addBlockRunner`. @bolt12 it's not about the thread itself, it's about all the other concurrent writes to the `ChainDB` that are done by all block-fetch clients.

@nfrisby The exception will propagate and eventually `ChainDB.closeDB` will be called.  It seems to me that we don't have access to the `TVar` which holds `ChainDbState` in the context of `addBlockRunner`, so keeping the state of the queue might be an easier option to implement (as in the suggestion you stroked).